File: Kotlin/Maths/TowerofHanoi.kt
/*
The Towers of Hanoi is a game where you have three pegs (#1, #2, and #3) and some circular disks of different sizes that slide onto the pegs.
They all start on one peg, largest to smallest (largest on the bottom). 
The goal is to move all the disks to another peg by following these rules:
you may only move one disk at a time from peg to peg no disk may be placed on top of a smaller disk.
    
Program will print the source and destination to transfer each disks for all the steps in chronological order.
*/

class Hanoi(disks: Int) {
    private var moves = 0

    init {
        println("Towers of Hanoi with $disks disks:\n")
        move(disks, 'L', 'C', 'R')                     
        println("\nCompleted in $moves moves\n")
    }

    private fun move(n: Int, from: Char, to: Char, via: Char) {
        if (n > 0) {
            move(n - 1, from, via, to)
            moves++
            println("Move disk $n from $from to $to")    
            move(n - 1, via, to, from)
        }
    }
}

fun main(args: Array<String>) {
     val sc = Scanner(System.`in`)

    println("Enter Number: ")                                
    val num: Int = sc.nextInt()
    Hanoi(num);
}

/*
Test Case :
******************************************************************
Input : 3
Output :
Towers of Hanoi with 3 disks:
Move disk 1 from L to C
Move disk 2 from L to R
Move disk 1 from C to R
Move disk 3 from L to C
Move disk 1 from R to L
Move disk 2 from R to C
Move disk 1 from L to C
******************************************************************
Time Complexity: O(2^n)
Space Complexity: O(n)
*/

File: Dart/maths/TowerOfHanoi.dart
/*
Tower of Hanoi is a mathematical puzzle where we have three rods and n disks. 
The objective of the puzzle is to move the entire stack to another rod, obeying the following simple rules:
1) Only one disk can be moved at a time.
2) Each move consists of taking the upper disk from one of the stacks and placing it on top of another stack i.e. a disk can only be moved if it is the uppermost disk on a stack.
3) No disk may be placed on top of a smaller disk.
*/

import 'dart:io';

void towerOfHanoi (int num, String from,String aux, String to){
  
  // Base case
  if (num == 1){
    print("Move Disk from Rod-$from to Rod-$to");
  }
  // Recursive case
  else{
    towerOfHanoi(num-1, from, to, aux);
    towerOfHanoi(1, from, aux, to);
    towerOfHanoi(num-1, aux, from, to);
  }
}

// Main function with driver function
void main(){
  // Take user input
  print("Enter Number of disk:");
  int num = int.parse(stdin.readLineSync()!);

  // Pass to Tower of Hanoi Recurrsive function
  towerOfHanoi(num, 'A','B', 'C');
}

/**
Time Complexity: O(2^N)
Space Complexity: O(N)

Sample input/output:

Enter Number of disk:
2
Move Disk from Rod-A -> Rod-B
Move Disk from Rod-A -> Rod-C
Move Disk from Rod-B -> Rod-C

Enter Number of disk:
3
Move Disk from Rod-A to Rod-C
Move Disk from Rod-A to Rod-B
Move Disk from Rod-C to Rod-B
Move Disk from Rod-A to Rod-C
Move Disk from Rod-B to Rod-A
Move Disk from Rod-B to Rod-C
Move Disk from Rod-A to Rod-C

*/
